### PR TITLE
ycsb: protect zetaN with the mutex

### PR DESCRIPTION
**Pull request by petermattis**
ZipfGenerator.Uint64 was accessing ZipfGenerator.zetaN without a lock
even though this field is updated concurrently by
ZipfGenerator.IncrementIMax.

Fix buglet in --initial-load. We were loading 1 too many keys initially,
causing the first write to hit a duplicate key constraint error.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/loadgen/31)

### CODE REVIEW

**Comment by petermattis** [comment]
@a-robinson I think @arjunravinarayan is on vacation. Noticed this during code inspection. I'm not sure how it could cause the `denim` perf oddities, but worth fixing regardless.

File: ycsb/main.go [hunk 1]
```diff
@@ -206,7 +206,7 @@ func (yw *ycsbWorker) runLoader(n uint64, numWorkers int, thisWorkerNum int, wg
 	if *verbose {
 		fmt.Printf("Worker %d loading %d rows of data\n", yw.workerID, n)
```
Comment (a-robinson) [inline ycsb/main.go]: Not your code, but this statement appears incorrect - worker `yw.workerID` is loading much less than `n` rows of data judging by the loop below

**Review by a-robinson — APPROVED** [review]
LGTM

File: ycsb/main.go (hunk 1 above)
Comment (petermattis) [reply ycsb/main.go]: I've removed this log message and a few others which weren't particularly useful.

**Comment by petermattis** [comment]
Updated to move `ZipfGenerator.r` under the mutex. PTAL.

File: ycsb/main.go [hunk 2]
```diff
@@ -466,9 +448,12 @@ func main() {
 
 	loadStart := time.Now()
 	var wg sync.WaitGroup
-	for i := range workers {
+	// TODO(peter): Using all of the workers for loading leads to errors with
+	// some of the insert statements receiving an EOF. For now, use a single
+	// worker.
+	for i, n := 0, 1; i < n; i++ {
```
Comment (a-robinson) [inline ycsb/main.go:454]: It's fine if we actually plan on coming back and changing it, but this is a really strange way of doing something once.

**Review by a-robinson — APPROVED** [review]
LGTM. Did this fix things?

File: ycsb/main.go (hunk 2 above)
Comment (petermattis) [reply ycsb/main.go:454]: I'm going to take a look at this tonight.

**Comment by petermattis** [comment]
> Did this fix things?

Yes. See the latest message in https://github.com/cockroachdb/cockroach/issues/13627.

**Comment by rjnn** [comment]
Oof, good catch.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/31#-:-Kf7K8MqHNJcjFfI3tDj:b199fy9)*

**Pull request merged**